src/finch/compiled.py

```diff
@@ -29,6 +34,45 @@ def _to_lazy_tensor(x: Tensor | Any, /) -> Tensor | Any:
     return x if not isinstance(x, Tensor) else lazy(x)
 
 
+def _recurse_iter_return(x: IterObj, /, *, compute_kwargs: dict[str, Any]) -> IterObj:
+    current = 0
+
+    def recursor(_: Any) -> tuple[int, Any]:
+        nonlocal current
+        ret = current
+        current += 1
+        return ret
```

Comment:
Just my preference, but I consider `global` and `nonlocal` rather obfuscating, and I think return type is incorrect here.
I think `current` and `recursor` names don't reflect the actual purpose they serve - we just want to record/visit all input args (regardless how nested they are). WDYT we have a simple ArgRecorder (or any better name) dataclass like this:
```python
@dataclass
class ArgRecorder:
    acc: int = 0
    def count(self, _) -> int:
        self.acc += 1
        return self.acc
idx_obj = _recurse(x, f=ArgRecorder().count)
```
This way we won't have the nonlocal `current = 0`.

Reply:
That's much simpler indeed! Done in https://github.com/finch-tensor/finch-tensor-python/pull/94/commits/61455d58946d2f5fe8858c39eb9e14b55b941adc.